### PR TITLE
fix embeded kernels on Windows

### PR DESCRIPTION
**Pull request by apwojcik**
The PR makes searching for embedded kernels possible on Windows.

### CODE REVIEW

**Comment by averinevg** [comment]
@atamazov Please review

**Review by DrizztDoUrden — APPROVED** [review]

**Comment by junliume** [comment]
@apwojcik we have some CI failures, could you help to take a look?

File: addkernels/addkernels.cpp [hunk 1]
```diff
@@ -47,10 +47,10 @@ void Bin2Hex(std::istream& source,
     if(variable.length() != 0)
     {
         target << "extern const size_t " << variable << "_SIZE;" << std::endl;
-        target << "extern const unsigned char " << variable << "[];" << std::endl;
+        target << "extern const char " << variable << "[];" << std::endl;
```
Comment (atamazov) [inline addkernels/addkernels.cpp:50]: [Q] Why this change is necessary?

File: src/comgr.cpp [hunk 2]
```diff
@@ -585,14 +585,14 @@ class Dataset : ComgrOwner
     ~Dataset() { EC(amd_comgr_destroy_data_set(handle)); }
     auto GetHandle() const { return handle; }
     void AddData(const Data& d) const { EC_THROW(amd_comgr_data_set_add(handle, d.GetHandle())); }
-    void AddData(const std::string& name,
-                 const std::string& content,
+    void AddData(const fs::path& name,
```
Comment (atamazov) [inline src/comgr.cpp]: [Q] Why this change is necessary? Formally speaking, COMgr has nothing to do with `fs:path` semantics. Is it designed to abstract us from compiler and filesystem. `amd_comgr_set_data_name` deals with C strings that work as "text labels".

File: src/comgr.cpp [hunk 3]
```diff
@@ -745,8 +745,8 @@ static inline bool IsWave64Enforced(const OptionList& opts)
         opts.begin(), opts.end(), [](const std::string& s) { return s == "-mwavefrontsize64"; });
 }
 
-void BuildHip(const std::string& name,
-              const std::string& text,
+void BuildHip(const fs::path& name,
```
Comment (atamazov) [inline src/comgr.cpp]: Ditto.

File: src/comgr.cpp [hunk 4]
```diff
@@ -871,8 +870,8 @@ void BuildHip(const std::string& name,
     }
 }
 
-void BuildOcl(const std::string& name,
-              const std::string& text,
+void BuildOcl(const fs::path& name,
```
Comment (atamazov) [inline src/comgr.cpp]: Same

File: src/comgr.cpp [hunk 5]
```diff
@@ -957,8 +956,8 @@ void BuildOcl(const std::string& name,
     }
 }
 
-void BuildAsm(const std::string& name,
-              const std::string& text,
+void BuildAsm(const fs::path& name,
```
Comment (atamazov) [inline src/comgr.cpp]: Same

File: src/comgr.cpp [hunk 6]
```diff
@@ -1055,27 +1054,23 @@ struct Error : std::exception
     throw Error{s, text};
 }
 
-static inline std::string to_string(const std::string& v) { return {v}; }
-static inline std::string to_string(const char* v) { return {v}; }
-static inline auto to_string(const std::size_t& v) { return std::to_string(v); }
```
Comment (atamazov) [inline src/comgr.cpp:1056]: Please revert this. `HIPRTC_CALL_BASE` macro accepts `info` of any type. I want `info` to have the `to_string()` converter as a spare level of protection.

File: src/comgr.cpp [hunk 7]
```diff
@@ -1055,27 +1054,23 @@ struct Error : std::exception
...
     return std::string(reason) + " (" + std::to_string(static_cast<int>(status)) + ')';
 }
 
-#define HIPRTC_CALL_BASE(call, info, action, statusdef)                                         \
-    do                                                                                          \
-    {                                                                                           \
-        statusdef status = (call);                                                              \
-        if(status != HIPRTC_SUCCESS)                                                            \
-        {                                                                                       \
-            MIOPEN_LOG_E("\'" #call "\' " << to_string(info) << ": " << GetStatusText(status)); \
```
Comment (atamazov) [inline src/comgr.cpp]: Ditto

File: src/comgr.cpp [hunk 8]
```diff
@@ -1055,27 +1054,23 @@ struct Error : std::exception
...
-            (action);                                                                           \
-        }                                                                                       \
-        else if(miopen::IsEnabled(MIOPEN_ENV(MIOPEN_DEBUG_COMGR_LOG_CALLS)))                    \
-            MIOPEN_LOG_I("Ok \'" #call "\' " << to_string(info));                               \
```
Comment (atamazov) [inline src/comgr.cpp]: Ditto.

File: src/comgr.cpp [hunk 9]
```diff
@@ -1104,16 +1099,18 @@ static void PrintVersion()
 static void hiprtc_program_destroy(hiprtcProgram prog) { hiprtcDestroyProgram(&prog); }
 using hiprtc_program_ptr = MIOPEN_MANAGE_PTR(hiprtcProgram, hiprtc_program_destroy);
 
-static hiprtc_program_ptr CreateProgram(const char* src,
-                                        const char* name,
+static hiprtc_program_ptr CreateProgram(const std::string_view src,
+                                        const fs::path& name,
```
Comment (atamazov) [inline src/comgr.cpp]: We should not have any filesystem related semantics here.

File: src/hip/hip_build_utils.cpp [hunk 10]
```diff
@@ -107,8 +107,8 @@ const char* getOffloadBundlerBinPath()
 
 namespace miopen {
 
-static fs::path HipBuildImpl(boost::optional<TmpDir>& tmp_dir,
-                             const std::string& filename,
+static fs::path HipBuildImpl(const TmpDir& tmp_dir,
+                             const fs::path& filename,
```
Comment (atamazov) [inline src/hip/hip_build_utils.cpp]: This `filename` is actually a text label that denotes that chunk of source code in our internal storage.

It is Ok to use this text label to construct fs::path objects, however.

File: src/hip/hip_build_utils.cpp [hunk 11]
```diff
@@ -209,8 +208,8 @@ static fs::path HipBuildImpl(boost::optional<TmpDir>& tmp_dir,
 #endif
 }
 
-fs::path HipBuild(boost::optional<TmpDir>& tmp_dir,
-                  const std::string& filename,
+fs::path HipBuild(const TmpDir& tmp_dir,
+                  const fs::path& filename,
```
Comment (atamazov) [inline src/hip/hip_build_utils.cpp]: Same

File: src/hipoc/hipoc_program.cpp [hunk 12]
```diff
@@ -203,32 +203,35 @@ HIPOCProgramImpl::HIPOCProgramImpl(const std::string& program_name,
 }
 
 #if !MIOPEN_USE_COMGR
-void HIPOCProgramImpl::BuildCodeObjectInFile(std::string& params,
-                                             const std::string& src,
-                                             const std::string& filename)
+void HIPOCProgramImpl::BuildCodeObjectInFile(std::string params,
+                                             std::string_view src,
+                                             const fs::path& filename)
```
Comment (atamazov) [inline src/hipoc/hipoc_program.cpp]: Same

**Review by atamazov — CHANGES_REQUESTED** [review]
Partial review. We shouldn't drag fs::path semantics into code that works with our "repository" of kernel sources.

**Comment by atamazov** [comment]
@junliume @apwojcik I highly recommend doing separate PRs for _fixes_ and _refactorings_.

File: src/comgr.cpp (hunk 6 above)
Comment (apwojcik) [reply src/comgr.cpp:1056]: The code formats a message using a stream object. That is much safer and more versatile than the `to_string()` functions mentioned. I see those functions as redundant. Can you please elaborate more on the level of protection you mentioned?

Comment (apwojcik) [reply src/comgr.cpp:1056]: I checked, and on Windows, we are getting here: `C:\...\MIOpen\src\comgr.cpp:1115:5: error: no matching function for call to 'to_string'`. That is the reason for the change.

Comment (apwojcik) [reply src/comgr.cpp:1056]: I can see it now. No need for additional explanation. I will revert it.

File: addkernels/addkernels.cpp (hunk 1 above)
Comment (apwojcik) [reply addkernels/addkernels.cpp:50]: It is incompatible with the `std::string`, `std::string_view`, and stream data types. Later in the original code, it is cast to `const char*` to instantiate a string object (see CMake generated code). It makes no difference for binary data converting to hex if the data type is signed or unsigned—the correct value will be written. Converting to `char` simplifies a lot.

**Comment by apwojcik** [comment]
> @junliume @apwojcik I highly recommend doing separate PRs for _fixes_ and _refactorings_.

The PR has shrunk to the minimum required.

File: addkernels/addkernels.cpp (hunk 1 above)
Comment (atamazov) [reply addkernels/addkernels.cpp:50]: [resolved]

File: src/comgr.cpp (hunk 6 above)
Comment (atamazov) [reply src/comgr.cpp:1056]: [resolved]

File: src/kernels/kernel_includes.cpp.in [hunk 13]
```diff
@@ -36,62 +37,36 @@ ${KERNELS_DECLS}
...
-    auto m = kernel_includes();
-    std::transform(m.begin(),
-                   m.end(),
-                   std::back_inserter(keys),
-                   [](decltype(m)::value_type const& pair) { return pair.first; });
-    return keys;
-}
-
-std::vector<std::string> GetHipKernelIncList()
+const std::vector<std::reference_wrapper<const fs::path>>& GetKernelIncList()
```
Comment (atamazov) [inline src/kernels/kernel_includes.cpp.in:59]: ⚓

Comment (atamazov) [inline src/kernels/kernel_includes.cpp.in:59]: > > @junliume @apwojcik I highly recommend doing separate PRs for _fixes_ and _refactorings_.
> 
> The PR has shrunk to the minimum required.

just FYI - I do not mind , but this is still refactoring (albeit the good one).

File: src/include/miopen/kernel.hpp [hunk 14]
```diff
@@ -26,17 +26,16 @@
...
 
 namespace miopen {
-std::string GetKernelSrc(std::string name);
-std::string GetKernelInc(std::string key);
-const std::string* GetKernelIncPtr(std::string key);
-std::vector<std::string> GetKernelIncList();
-std::vector<std::string> GetHipKernelIncList();
+std::string_view GetKernelSrc(const fs::path& name);
+std::string_view GetKernelInc(const fs::path& name);
+const std::vector<std::reference_wrapper<const fs::path>>& GetKernelIncList();
```
Comment (atamazov) [inline src/include/miopen/kernel.hpp:38]: If we are going to keep `fs::path` as a "keys" in the "database of kernel sources" then I so not see any good in resolving comments from my previous review. Maybe it would be even better to keep using fs:path in comgr etc.

I know that @DrizztDoUrden is thinking that `fs:path` is good, and I am not going to argue endlessly. And finally you have my "minority report".

**Review by atamazov — APPROVED** [review]
No objections.

File: src/comgr.cpp (hunk 2 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved][I do not care]

File: src/comgr.cpp (hunk 3 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved][I do not care]

File: src/comgr.cpp (hunk 4 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved][I do not care]

File: src/comgr.cpp (hunk 5 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved][I do not care]

File: src/comgr.cpp (hunk 7 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved]

File: src/comgr.cpp (hunk 8 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved]

File: src/comgr.cpp (hunk 9 above)
Comment (atamazov) [reply src/comgr.cpp]: [resolved][I do not care]

File: src/hip/hip_build_utils.cpp (hunk 10 above)
Comment (atamazov) [reply src/hip/hip_build_utils.cpp]: [resolved][I do not care]

File: src/hip/hip_build_utils.cpp (hunk 11 above)
Comment (atamazov) [reply src/hip/hip_build_utils.cpp]: [resolved][I do not care]

File: src/hipoc/hipoc_program.cpp (hunk 12 above)
Comment (atamazov) [reply src/hipoc/hipoc_program.cpp]: [resolved][I do not care]

**Review by junliume — APPROVED** [review]
